### PR TITLE
Added documentation

### PR DESCRIPTION
**Pull request by sthirugn**

### CODE REVIEW

**Comment by omaciel** [comment]
I have made some suggestions to @sthirugn in private and have asked him to incorporate some of it to this document. If you want to merge it as is and then add my suggestions, I'm fine with it too.

File: .gitignore [hunk 1]
```diff
@@ -55,3 +55,7 @@ docs/_build/
 
 # PyBuilder
 target/
+
+# Misc
```
Comment (elyezer) [inline .gitignore]: These can be placed outside the betelgeuse repo, since it is not related to the project is better to remove the ignore and use on the home for .pylarion and clone pylarion elsewhere.

File: data_sample/tests/test_login_example.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+"""Test class for Login"""
+
+
+class LoginTestCase():
+    """Tests for Login"""
+
+    def test_login_1(self):
+        """Check if a user is able to login with valid userid and password
+
+        @Feature: Login
```
Comment (elyezer) [inline data_sample/tests/test_login_example.py]: Feature is not a token that betelgeuse looks, use requirement instead.

File: data_sample/tests/test_login_example.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
+# -*- encoding: utf-8 -*-
```
Comment (elyezer) [inline data_sample/tests/test_login_example.py]: `-*- encoding: utf-8 -*-` is a format for emacs python default one is `coding=utf-8` better to use it.

For more information check https://www.python.org/dev/peps/pep-0263/.

File: data_sample/tests/test_login_example.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    """Tests for Login"""
+
+    def test_login_1(self):
+        """Check if a user is able to login with valid userid and password
+
+        @Feature: Login
+
+        @id: 60e48736-43a9-11e6-bcaa-104a7da122d7
+
+        @Assert: User is able to login successfully
```
Comment (elyezer) [inline data_sample/tests/test_login_example.py]: Assert is not a token betelgeuse looks, use expectedresults instead.

File: data_sample/tests/test_login_example.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+# -*- encoding: utf-8 -*-
+"""Test class for Login"""
+
+
+class LoginTestCase():
```
Comment (elyezer) [inline data_sample/tests/test_login_example.py]: Since this is showing an example of a class which probably will be a unittest TestCase child is better to show it here. Otherwise a plain test function (test_loging_1) can be used.

File: docs/index.rst [hunk 6]
```diff
@@ -1,14 +1,162 @@
...
+
+What is Betelgeuse?
+```````````````````
+
+Betelgeuse is a python program that reads standard Python test cases and offers
+tools to interact with Polarion. Possible interactions are:
+
+* Automatic creation/updation of Requirements and Test Cases from a Python
   project code base.
 * Creation of Test Runs based on a jUnit XML file.
```
Comment (elyezer) [inline docs/index.rst]: This should also state that it needs the source code as well in order to proper map the ID.

File: docs/index.rst [hunk 7]
```diff
@@ -1,14 +1,162 @@
...
 
-Documentation contents:
+Betelgeuse uses `testimony <https://pypi.python.org/pypi/testimony>`_ and
```
Comment (elyezer) [inline docs/index.rst]: Pylarion is not used to parse testcase. It is a client for the Polarion WSDL API. Will be more clear to change this to:

``` rst
Betelgeuse uses `testimony <https://pypi.python.org/pypi/testimony>`_ to parse test cases and Pylarion to talk with the Polarion's WSDL API.
```

File: docs/index.rst [hunk 8]
```diff
@@ -1,14 +1,162 @@
...
+respectively.
+
+.. _prerequisites:
+
+Prerequisites
+`````````````
+Login to polarion in a browser to check if the login user has permissions to do
+the following:
+
+* add/update test case
```
Comment (elyezer) [inline docs/index.rst]: s/add/create let's normalize it and the line below.

File: docs/index.rst [hunk 9]
```diff
@@ -1,14 +1,162 @@
...
+* add/update requirements
+* create test runs
```
Comment (elyezer) [inline docs/index.rst]: create/update test runs

Betelgeuse is able to update a previously created test run and will add just the results that is not already on the test run.

File: docs/index.rst [hunk 10]
```diff
@@ -1,14 +1,162 @@
...
+1. betelgeuse uses pylarion to interact with polarion. Install pylarion from its
+   source.
+
+   .. note::
+
+     - It may be possible that the latest version of betelgeuse may not work
+       correctly with some versions of pylarion.  In this case, please use an
+       alternate working version of pylarion.
+     - Read pylarion documentation and set up .pylarion config file as
+       required.
```
Comment (elyezer) [inline docs/index.rst:46]: A sample .pylarion file can be added here:

```
[webservice]
url=https://<polarion url>
svn_repo=https://<repo url>
user=<username>
password=<password>
default_project=<default_project>
```

Also worth mention that this file should be place at the user's home directory.

File: docs/index.rst [hunk 11]
```diff
@@ -1,14 +1,162 @@
...
+       $ pip install betelgeuse
+
+3. Alternatively you can fork betelgeuse to your github account and clone it
+   locally.  And then you can install it locally by:
+
+   .. code-block:: bash
+
+       $ cd betelgeuse
+       $ pip install -r requirements.txt
+       $ pip install -e .
```
Comment (elyezer) [inline docs/index.rst:60]: requirements.txt already install Betelgeuse in editable mode. This line can be removed.

File: docs/index.rst [hunk 12]
```diff
@@ -1,14 +1,162 @@
...
+   .. code-block:: bash
+
+       $ pip install betelgeuse
+
+3. Alternatively you can fork betelgeuse to your github account and clone it
+   locally.  And then you can install it locally by:
+
+   .. code-block:: bash
+
+       $ cd betelgeuse
```
Comment (elyezer) [inline docs/index.rst:59]: Add a line above with `git clone https://github.com/SatelliteQE/betelgeuse.git`

File: docs/index.rst [hunk 13]
```diff
@@ -1,14 +1,162 @@
...
+   .. note:: It is always recommended to use python virtual environment
+
+.. _usage_examples:
+
+Usage Examples
+``````````````
+.. note::
+
+  For easy understanding of betelgeuse, this repository is already included with
+  `data_sample` folder.  This folder contains sample tests and xml results which
```
Comment (elyezer) [inline docs/index.rst]: better to name sample_project specially because both Betelgeuse and the test runner should be run on the same directory, this will ensure the test run ID mapping will work.

File: docs/index.rst [hunk 14]
```diff
@@ -1,14 +1,162 @@
...
+
+.. code-block:: bash
+
+    $ betelgeuse --help
+
+.. _test_case_command:
+
+test-case command
++++++++++++++++++
+Creates/Syncs test cases with polarion. This command performs the following
```
Comment (elyezer) [inline docs/index.rst]: Create/update test cases on Polarion...

File: docs/index.rst [hunk 15]
```diff
@@ -1,14 +1,162 @@
...
+steps:
+
+- Testimony is called to parse the test cases.
+- For each parsed test case, the following actions are performed:
+
+    - The requirement name is derived from the test module name.  If the test
```
Comment (elyezer) [inline docs/index.rst]: It will first check if the test method have a `@Requirement` token and create/fetch the requirement after it. If that is not present then it will default to the module name.

File: docs/index.rst [hunk 16]
```diff
@@ -1,14 +1,162 @@
...
+      module name is `test_login_example.py`, then the requirement name is
+      `Login Example`.
```
Comment (elyezer) [inline docs/index.rst]: Inline code on RST is `content`.

File: docs/index.rst [hunk 17]
```diff
@@ -1,14 +1,162 @@
...
+    - Requirement is created.
+    - Test case is created and linked to the requirement.
```
Comment (elyezer) [inline docs/index.rst]: This should be expanded saying about the default values and the expected tokens.

File: docs/index.rst [hunk 18]
```diff
@@ -1,14 +1,162 @@
...
+
+.. code-block:: bash
+
+    $ betelgeuse test-case --path data_sample/tests/ PROJECT_CLOUD
```
Comment (elyezer) [inline docs/index.rst]: I would just add the command call without the output, the output is easier to get outdated

File: docs/index.rst [hunk 19]
```diff
@@ -1,14 +1,162 @@
...
+Creates a test run in polarion using the information in the given junit xml
+file. This command performs the following steps:
+
+- Parses the junit xml file to read all the test cases and their run statuses.
+- Creates a new test run and updates it with all the parsed test case items and
+  their run statuses.
+
+.. code-block:: bash
+
+    $ betelgeuse test-run --path data_sample/results/sample-junit-result.xml \
```
Comment (elyezer) [inline docs/index.rst]: In order to Betelgeuse proper map the `@ID` with the jUnit result it is needed that both the test runner and Betelgeuse get called on the same directory. With that said by having an entry like the following on the jUnit XML report:

``` xml
<testcase classname="data_sample.tests.test_login_example.LoginTestCase" file="data_sample/tests/test_login_example.py" line="421" name="test_login_1" time="694.768339396">...</testcase>
```

Betelgeuse will map the test method based on its classname and name by joining them with a dot: `data_sample.tests.test_login_example.LoginTestCase.test_login_1` then Betelgeuse will convert every test module path into a python import path for example `data_sample/tests/test_login_example.py` will become `data_sample.tests.test_login_example` and all test methods or functions will then be appended later on, for the method `test_login_1` inside the class `LoginTestCase` it will generate `data_sample.tests.test_login_example.LoginTestCase.test_login_1`. With the two information matching then it will check the `@ID` token of that test method or function and will query Polarion for the work item id that represents that ID, finally with the work_item_id information Betelgeuse will add the result for a Polarion test case (work item) into the test run.

It is important to note that if Betelgeuse can't find the `@ID` token information the ID will default to the Python import path, in the previous example it will be `data_sample.tests.test_login_example.LoginTestCase.test_login_1`.

If no result is returned when querying Polarion for a matching test case, that result will be skipped and the next entry on the jUnit will be processed. It is highly recommended to run test-case command before test-run to make sure all test cases are created/updated.

**Comment by elyezer** [comment]
Please normalize all betelgeuse in lower case to Betelgeuse. The same can be done for Testimony and Polarion. Also jUnit XML is how Betelgeuse help message reports about the jUnit XML report.

File: docs/index.rst [hunk 20]
```diff
@@ -1,14 +1,162 @@
...
+.. _quick_start:
+
+Quick Start
+```````````
+1. betelgeuse uses pylarion to interact with polarion. Install pylarion from its
+   source.
+
+   .. note::
+
+     - It may be possible that the latest version of betelgeuse may not work
```
Comment (elyezer) [inline docs/index.rst]: Currently using the alternate Pylarion version is a must since it adds ability to compare TestSteps which is needed in order to proper populate test steps for the test case.

**Comment by sthirugn** [comment]
Thanks for all the valuable comments.  I will work on each of these and respond back.

File: .gitignore (hunk 1 above)
Comment (sthirugn) [inline .gitignore]: For me as a developer, pylarion is used only with betelgeuse and I like to keep betelgeuse and pylarion in one folder for my easy maintenance.  I thought this would help other developers too.  Having said that, I am happy to move this to my general .gitignore if this is not going to help fellow developers.

File: docs/index.rst (hunk 6 above)
Comment (sthirugn) [inline docs/index.rst]: `Automatic creation/updation of Requirements and Test Cases from a Python project code base.` implies that this tool needs source code access.  The requirement creation mechanism is also detailed later in the doc which should cover this.

File: docs/index.rst (hunk 7 above)
Comment (sthirugn) [inline docs/index.rst]: I have used the `respectively` construct to indicate that `testimony is used to parse test cases` and `pylarion is used to communicate with polarion`. 

Reference: https://www.quora.com/How-can-we-use-respectively-in-a-sentence-And-what-is-the-meaning-of-respectively-in-a-sentence

File: docs/index.rst (hunk 8 above)
Comment (sthirugn) [inline docs/index.rst]: Sure I can update `create` instead of `add`.  What do you mean by `normalize` here?

File: docs/index.rst (hunk 20 above)
Comment (sthirugn) [inline docs/index.rst]: I have been told that some teams use original pylarion and other teams using forked version.  So this is a note to the user that he should be mindful of using the right pylarion version.  Also I am refraining from adding too much info on pylarion here as this is an internal project.

In the long run, hopefully our fixes get accepted in original pylarion and we don't need to fork it anymore.  So this is a kind of temporary solution.

File: docs/index.rst (hunk 10 above)
Comment (sthirugn) [inline docs/index.rst:46]: I respectfully disagree. I like to keep betelgeuse project as atomic as possible.  This doc already says above that the user should read pylarion documentation to set .pylarion config file.  Any sample .pylarion file should live in the pylarion project README in my opinion.

Also pylarion documentation clearly says that the config file can live in current directory, user's home directory, or /etc/pylarion/pylarion.cfg.  I don't think we need to restrict betelgeuse users to keep the config in home directory only.

File: docs/index.rst (hunk 13 above)
Comment (sthirugn) [inline docs/index.rst]: Good point.  I can 
- leave the name as `data_sample`, if that is okay.
- add a note to say that the test runner and betelgeuse should be run on the same directory

File: docs/index.rst (hunk 17 above)
Comment (sthirugn) [inline docs/index.rst]: Currently I am not much familiar about the betelgeuse project code, I can improve this once I start using betelgeuse more.

File: docs/index.rst (hunk 18 above)
Comment (sthirugn) [inline docs/index.rst]: I have seen many open source projects in which the output was shown and it gave me confidence that I am doing the steps correctly.  I am following the same logic here to help any future betelgeuse learner.

Also I think it is the responsibility of the developer to keep this up to date, if we need to attract more new users.

File: .gitignore (hunk 1 above)
Comment (elyezer) [inline .gitignore]: As you can we have two different uses, it is better to leave each developer add its own ignores outside.

File: docs/index.rst (hunk 7 above)
Comment (elyezer) [inline docs/index.rst]: Never mind, I didn't saw the respectively on line 20. It was too late when I review this, I'm sorry.

File: docs/index.rst (hunk 8 above)
Comment (elyezer) [inline docs/index.rst]: I mean to normalize all terms to the same like using create and update everywhere.

File: docs/index.rst (hunk 20 above)
Comment (elyezer) [inline docs/index.rst]: Agreed, we can provide the detailed information on the right channels other than here.

File: docs/index.rst (hunk 17 above)
Comment (sthirugn) [inline docs/index.rst]: This will be addressed in https://github.com/SatelliteQE/betelgeuse/issues/55

File: docs/index.rst (hunk 10 above)
Comment (elyezer) [inline docs/index.rst:46]: I suggested adding it here just as quick setup information, always consulting the official Pylarion documentation will be needed for more detailed information.

Feel free to add or not the the .pylarion sample here.

File: docs/index.rst (hunk 13 above)
Comment (elyezer) [inline docs/index.rst]: I suggested to change the name because the word data is too generic and we are talking about a Python project, if we can mimic a real Python project with a clear name I think that would be better.

File: docs/index.rst (hunk 17 above)
Comment (elyezer) [inline docs/index.rst]: Fair enough. I can add that information later.

File: docs/index.rst (hunk 18 above)
Comment (elyezer) [inline docs/index.rst]: Let's have this in mind when reviewing PRs. The review process for now on should strictly ensure updated documentation.

**Comment by elyezer** [comment]
@sthirugn thanks for the follow up and great documentation improvements. I have left some responses and observations about your points, really good points by the way.

File: docs/index.rst (hunk 7 above)
Comment (sthirugn) [inline docs/index.rst]: No worries at all.

File: docs/index.rst (hunk 8 above)
Comment (sthirugn) [inline docs/index.rst]: ok I think you mean `make it consistent`.  Please check the updated commit and let me know what do you think.

File: docs/index.rst (hunk 10 above)
Comment (sthirugn) [inline docs/index.rst:46]: Thank you for the response.  I think we can add it in future, as needed.  Having said this, we will have to improve Pylarion and testimony docs up-to-date too! This will help in easy reference of docs among the projects instead of duplicating the same info in multiple places.

File: docs/index.rst (hunk 13 above)
Comment (sthirugn) [inline docs/index.rst]: I see your point and I think it will add value. It is updated now.

File: docs/index.rst (hunk 19 above)
Comment (sthirugn) [inline docs/index.rst]: This is a great explanation of the code logic.  I added this in docs now.

**Comment by sthirugn** [comment]
I believe I addressed all comments except for the open issues which will be addressed later.  Please let me know if I missed any comment.
- https://github.com/SatelliteQE/betelgeuse/issues/55
- https://github.com/SatelliteQE/betelgeuse/issues/56

File: docs/index.rst [hunk 21]
```diff
@@ -1,14 +1,207 @@
...
-* Automatic creation and update of Requirements and Test Cases from a Python
+.. _what_is_betelgeuse:
+
+What is Betelgeuse?
+```````````````````
+
+Betelgeuse is a python program that reads standard Python test cases and offers
+tools to interact with Polarion. Possible interactions are:
+
+* Automatic creation/updation of Requirements and Test Cases from a Python
```
Comment (omaciel) [inline docs/index.rst]: s/updation/update

File: docs/index.rst [hunk 22]
```diff
@@ -1,14 +1,207 @@
...
   project code base.
-* Creation of Test Runs based on a jUnit XML file.
+* Automatic Creation/updation of Test Runs based on a jUnit XML file.
```
Comment (omaciel) [inline docs/index.rst]: s/updation/update

Comment (elyezer) [inline docs/index.rst]: s/Creation/creation

File: docs/index.rst [hunk 23]
```diff
@@ -1,14 +1,207 @@
...
+.. note::
+
+  1. For easy understanding of Betelgeuse, this repository is already included with
+  ``sample_project`` folder.  This folder contains sample tests and XML results which
+  will help in setting up and testing Betelgeuse for your project.  The sample
+  commands used below also use this data.
+
+  2. Always run the test runner and Betelgeuse on the same directory to make
+  sure that the test run ID mapping works fine.  Otherwise Betelgeuse may
+  report ID errors.  More info in test_run_command_ section
```
Comment (omaciel) [inline docs/index.rst]: More info can be found in...

File: docs/index.rst [hunk 24]
```diff
@@ -1,14 +1,207 @@
...
+steps:
+
+- Testimony is called to parse the test cases.
+- For each parsed test case, the following actions are performed:
+
+    - If ``@requirement`` token is present in the test case, it will be used as the
+      requirement name.  Otherwise it is derived from the test module name.  For
+      example, if the test module name is ``test_login_example``, then the
+      requirement name is ``Login Example``.
+    - Requirement is created.
```
Comment (omaciel) [inline docs/index.rst]: First, we try to find the requirement by its name. If nothing is found, then we try to create one.

File: docs/index.rst [hunk 25]
```diff
@@ -1,14 +1,207 @@
...
+    - Test case is created and linked to the requirement.
```
Comment (omaciel) [inline docs/index.rst]: ... or updated.

File: docs/index.rst [hunk 26]
```diff
@@ -1,14 +1,207 @@
...
+
+.. note::
+  
+  * ``PROJECT_CLOUD`` is the polarion project id and not the project name.  This
+    can be found in Polarion -> Settings (icon) -> Administration -> ID.
+  * ``path`` is the path of the folder which has the test cases source code.
+
+.. warning::
+
+   Are you not sure if you are using this command in a right way?  No
```
Comment (omaciel) [inline docs/index.rst]: s/in a right way/correctly

File: docs/index.rst [hunk 27]
```diff
@@ -1,14 +1,207 @@
...
+
+.. _test_run_command:
+
+test-run command
+++++++++++++++++
+Creates a test run in polarion using the information in the given jUnit XML
+file. This command performs the following steps:
+
+- Parses the jUnit XML file to read all the test cases and their run statuses.
+- Creates a new test run and updates it with all the parsed test case items and
```
Comment (omaciel) [inline docs/index.rst]: iirc it will only create the test run if it cannot find one by name?

File: docs/index.rst [hunk 28]
```diff
@@ -1,14 +1,207 @@
...
+
+.. code-block:: xml
+
+    <testcase classname="sample_project.tests.test_login_example.LoginTestCase"
+    file="sample_project/tests/test_login_example.py" line="421" name="test_login_1"
+    time="694.768339396">...</testcase>
+
+With the above report, Betelgeuse performs the following:
+
+- Derives the test method by joining its ``classname`` and ``name`` attributes
```
Comment (omaciel) [inline docs/index.rst]: Derives the test method.. name? pseudo-id?

File: docs/index.rst [hunk 29]
```diff
@@ -1,14 +1,207 @@
...
+  with a dot.  In this case, it becomes 
+  ``sample_project.tests.test_login_example.LoginTestCase.test_login_1``.
+- Looks at the ``--source-code-path`` option value and converts every test
+  module path into a python import path.  For example,
+  ``sample_project/tests/test_login_example.py`` will become
+  ``sample_project.tests.test_login_example``.  All test methods or functions are
```
Comment (omaciel) [inline docs/index.rst]: Should `All test methods...` be a separate bullet item on the same level?

File: docs/index.rst [hunk 30]
```diff
@@ -1,14 +1,207 @@
...
+  then appended.  For example, the method ``test_login_1`` inside the class
+  ``LoginTestCase`` will be generated as
+  ``sample_project.tests.test_login_example.LoginTestCase.test_login_1``.
+- The information obtained from both the steps above are compared and ``@ID``
```
Comment (omaciel) [inline docs/index.rst:201]: I find this part a bit confusing and hard to follow (for anyone who's not familiar with the source code or Betelgeuse). Can we file an issue to improve it a bit more?

**Comment by omaciel** [comment]
@sthirugn thank you for updating this. I have added a few more nitpicks which I hope will make this first draft easier to understand and follow.

File: docs/index.rst (hunk 30 above)
Comment (sthirugn) [inline docs/index.rst:201]: This is explaining the code logic in the same way as it is written.  Assumption is that this needs an in-depth understanding of how betelgeuse works.  Can you type some examples of how you think this can be improved?

**Comment by sthirugn** [comment]
Ready for review again.

**Comment by omaciel** [comment]
ACK

**Comment by elyezer** [comment]
@sthirugn thank you very much! This is awesome.
